Core/Object Arts/Dolphin/IDE/Base/SmalltalkWorkspace.cls

```diff
@@ -902,46 +909,20 @@ messagesForToken: anAssociation startingWith: aString maxItems: anInteger
 	"Private - Answer the set of selectors that could potentially be sent as messages following
 	the specified token."
 
-	| ignoreCase icon |
+	| selectors class ignoreCase icon |
 	ignoreCase := self isAutoCompletionCaseInsensitive.
 	icon := Message icon.
-	^((self classForToken: anAssociation)
-		ifNil: 
-			[| allSelectors |
-			allSelectors := Smalltalk developmentSystem allSelectors.
-			aString isEmpty
-				ifTrue: [allSelectors size > anInteger ifTrue: [#()] ifFalse: [allSelectors]]
-				ifFalse: 
-					[| messages |
-					messages := Set new.
-					allSelectors do: 
-							[:eachName |
-							(eachName beginsWith: aString ignoreCase: ignoreCase)
-								ifTrue: 
-									[messages add: eachName.
-									messages size > anInteger ifTrue: [^#()]]].
-					messages]]
-		ifNotNil: 
-			[:class |
-			| messages |
-			messages := Set new.
-			class withAllSuperclassesDo: (aString isEmpty
-						ifTrue: 
-							[
-							[:eachClass |
-							messages addAll: eachClass selectors.
-							messages size > anInteger ifTrue: [^#()]]]
-						ifFalse: 
-							[
-							[:eachClass |
-							eachClass selectors do: 
-									[:eachSelector |
-									(eachSelector beginsWith: aString ignoreCase: ignoreCase)
-										ifTrue: 
-											[messages add: eachSelector.
-											messages size > anInteger ifTrue: [^#()]]]]]).
-			messages])
-			collect: [:each | ScintillaListItem text: each icon: icon]!
+	class := self classForToken: anAssociation.
+	class == UndefinedObject ifTrue: [class := nil].
+	selectors := Set new.
+	Smalltalk developmentSystem allSelectors do: 
```

Comment:
There is a (subtle but important) change in behavior, which in turn is the motivation for the change. The new code is willing to consider selectors that the class does not in fact "implement" in the usual sense, so long as it claims that it `#canUnderstand:` them (presumably via a `#doesNotUnderstand:` implementation). I realize this does depend on those selectors being present in the `allSelectors` collection in the first place—I've added a commit changing `#allSelectors` to use `#understoodSelectors`, giving such classes a chance to ensure this is the case.

Regarding the performance cases—all true. I guess the bottom line for me is that:
1. The worst-case is no worse than before—completing against a receiver of unknown type already examined everything, and is as you say the common case anyway.
2. Performance of code triggered only by user interaction (not potentially run in a tight loop) is just so much less of an issue than it was 10 years ago that I think the simplicity and ease of understanding is worth the performance cost, such as it is.